servers/rendering/renderer_rd/effects/metal_fx.h

```diff
@@ -0,0 +1,145 @@
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef METAL_FX_RD_H
+#define METAL_FX_RD_H
+
+#ifdef METAL_ENABLED
+
+#include "servers/rendering/renderer_scene_render.h"
+
+#ifdef __OBJC__
+@protocol MTLFXSpatialScaler;
+@protocol MTLFXTemporalScaler;
+#endif
+
+namespace RendererRD {
+
+struct MFXSpatialContext {
+#ifdef __OBJC__
+	id<MTLFXSpatialScaler> scaler = nullptr;
+#else
+	void *scaler = nullptr;
+#endif
+	MFXSpatialContext() = default;
+	~MFXSpatialContext();
+};
+
+class MFXSpatialEffect {
+	static const size_t RID_COUNT = 3;
+
+	struct CallbackArgs {
+		RDD::TextureID src;
+		RDD::TextureID dst;
+		MFXSpatialContext ctx;
+	};
+
+	uint32_t i = 0;
+	CallbackArgs args[RID_COUNT];
+	static void callback(RDD *p_driver, RDD::CommandBufferID p_command_buffer, CallbackArgs *p_userdata);
+
+public:
+	MFXSpatialEffect();
+	~MFXSpatialEffect();
+
+	struct CreateParams {
+		Vector2i input_size;
+		Vector2i output_size;
+		RD::DataFormat input_format;
+		RD::DataFormat output_format;
+	};
+
+	MFXSpatialContext *create_context(CreateParams p_params) const;
+	void process(MFXSpatialContext *p_ctx, RID p_src, RID p_dst);
+};
+
+struct MFXTemporalContext {
+#ifdef __OBJC__
+	id<MTLFXTemporalScaler> scaler = nullptr;
+#else
+	void *scaler = nullptr;
+#endif
+	MFXTemporalContext() = default;
+	~MFXTemporalContext();
+};
+
+class MFXTemporalEffect {
+	static const size_t RID_COUNT = 3;
```

Comment:
While inefficient, it'd be perfectly possible for a user to run into problems with this limitation, as just making four viewports that use scaling (e.g. player split-screen) would easily cause it. There's a few alternate solutions that'd allow this to scale without running into this issue that might be hard to catch in the future.

One possibility is that it should be pretty safe to assume that an argument structure can be discarded and reused once the callback for it is called. A growing vector and a free list should make it easy to achieve this.

Reply:
Right, good catch! Do we have existing examples or data structures for managing a free list or will I have to implement it myself?

Reply:
No ready-made template as far as I know, but the logic is pretty straightforward. Just keep two vectors and use one to track the indices that are free (free list) and the other one to track the contents. Whenever you try to use an element, check if there's anything in the free list. If not, expand the vector. On deletion, insert the index into the free list.

Reply:
Sounds good – I will sort that out.

Reply:
Done – using PagedAllocator with 16 elements, as that should be more than enough. It will allocate new pages of 16 elements if we do happen to go over.